File: day08/handheld.py
with open("input.txt") as f:
    instructions = [line.strip().split(" ") for line in f.readlines()]


def run_program(instr):
    indices = set()
    accumulator, index = 0, 0
    while True:
        instruction = instr[index][0]
        argument = int(instr[index][1])
        if instruction == "acc":
            accumulator += argument
            index += 1
        elif instruction == "jmp":
            index += argument
        elif instruction == "nop":
            index += 1
        if index >= len(instr):
            return 0, accumulator
        if index in indices:
            return 1, accumulator
        indices.add(index)


print(run_program(instructions)[1])


def fix_program(instr):
    for i, op in enumerate(instr):
        changed_instructions = instr.copy()
        if op[0] == "jmp":
            changed_instructions[i] = ["nop", op[1]]
        elif op[0] == "nop":
            changed_instructions[i] = ["jmp", op[1]]
        else:
            continue
        answer = run_program(changed_instructions)
        if answer[0] == 0:
            return answer[1]


print(fix_program(instructions))

File: day08/Handheld.go
package main

import (
	"bufio"
	"log"
	"os"
	"strconv"
	"strings"
)

type Instruction struct {
	Operation string
	Argument  int
}

func main() {
	var instructions []Instruction
	for _, line := range readInput() {
		split := strings.Split(line, " ")
		argument, _ := strconv.Atoi(split[1])
		instructions = append(instructions, Instruction{split[0], argument})
	}
	_, part1 := runProgram(instructions)
	println(part1)
	println(fixProgram(instructions))
}

func runProgram(instructions []Instruction) (int, int) {
	accumulator := 0
	seen := make(map[int]bool)
	for i := 0; i <= len(instructions); i++ {
		if i == len(instructions) {
			return 0, accumulator
		} else if seen[i] {
			return 1, accumulator
		}
		seen[i] = true
		instruction := instructions[i]
		switch instruction.Operation {
		case "acc":
			accumulator += instruction.Argument
		case "jmp":
			i += instruction.Argument - 1
		}
	}
	return 1, accumulator
}

func fixProgram(instructions []Instruction) int {
	for index, instruction := range instructions {
		changed := make([]Instruction, len(instructions))
		switch instruction.Operation {
		case "jmp":
			copy(changed, instructions)
			changed[index].Operation = "nop"
		case "nop":
			copy(changed, instructions)
			changed[index].Operation = "jmp"
		default:
			continue
		}
		exitCode, result := runProgram(changed)
		if exitCode == 0 {
			return result
		}
	}
	return 0
}

func readInput() []string {
	file, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	var data []string

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		data = append(data, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	return data
}
